### ISSUE TITLE
Daemonset should support `Rereate` Update Strategy

### ISSUE THREAD

**Issue by gyliu513** [kubernetes/kubernetes#52973]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug

/kind feature

**What happened**:

If I create a daemonset with hostNetwork, then when upgrade the daemonset, there will be some port conflict.

We should support `Recreate` update strategy for daemonset like deployment.

/cc @janetkuo 

/sig apps

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by roffe** [comment]
I have tried this on 1.7.6 with a nginx ingress daemonset but cannot  reproduce the bug you are describing.
```
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: nginx-ingress-controller
  namespace: ingress-controller
spec:
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 1
  template:
    metadata:
      labels:
        name: nginx-ingress-controller
    spec:
      serviceAccountName: ingress-controller
      nodeSelector:
        ingress: external
      hostNetwork: true
      containers:
      - image: nginxdemos/nginx-ingress:1.0.0-alpine
        imagePullPolicy: Always
        name: nginx-ingress
        ports:
        - containerPort: 80
          hostPort: 80
        - containerPort: 443
          hostPort: 443
        env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        # Uncomment the lines below to enable extensive logging and/or customization of
        # NGINX configuration with configmaps
        args:
         #- -v=3
          - -nginx-configmaps=$(POD_NAMESPACE)/nginx-config
          - -default-server-tls-secret=$(POD_NAMESPACE)/default-server-secret
```

When I make changes & apply my daemonset I can, using a watch see that it terminates one pod, then brings a new one up. No port conflicts happening.

```
➜  nginx-ingress git:(master) ✗ kc get po --namespace ingress-controller --watch
NAME                             READY     STATUS    RESTARTS   AGE
nginx-ingress-controller-94tvx   1/1       Running   1          3d
nginx-ingress-controller-hx566   1/1       Running   1          3d
nginx-ingress-controller-p7cc5   1/1       Running   1          3d 
<-- Here i applied my new daemonset template -->
nginx-ingress-controller-hx566   1/1       Terminating   1         3d
nginx-ingress-controller-hx566   0/1       Terminating   1         3d
nginx-ingress-controller-hx566   0/1       Terminating   1         3d
nginx-ingress-controller-hx566   0/1       Terminating   1         3d
nginx-ingress-controller-hx566   0/1       Terminating   1         3d
nginx-ingress-controller-5n751   0/1       Pending   0         0s
nginx-ingress-controller-5n751   0/1       ContainerCreating   0         0s
nginx-ingress-controller-5n751   1/1       Running   0         9s
nginx-ingress-controller-94tvx   1/1       Terminating   1         3d
nginx-ingress-controller-94tvx   0/1       Terminating   1         3d
nginx-ingress-controller-5n751   1/1       Running   0         10s
nginx-ingress-controller-94tvx   0/1       Terminating   1         3d
nginx-ingress-controller-94tvx   0/1       Terminating   1         3d
nginx-ingress-controller-94tvx   0/1       Terminating   1         3d
nginx-ingress-controller-f75jr   0/1       Pending   0         0s
nginx-ingress-controller-f75jr   0/1       ContainerCreating   0         0s
nginx-ingress-controller-f75jr   1/1       Running   0         5s
nginx-ingress-controller-p7cc5   1/1       Terminating   1         3d
nginx-ingress-controller-p7cc5   0/1       Terminating   1         3d
nginx-ingress-controller-f75jr   1/1       Running   0         6s
nginx-ingress-controller-p7cc5   0/1       Terminating   1         3d
nginx-ingress-controller-p7cc5   0/1       Terminating   1         3d
nginx-ingress-controller-p7cc5   0/1       Terminating   1         3d
nginx-ingress-controller-pnch3   0/1       Pending   0         0s
nginx-ingress-controller-pnch3   0/1       ContainerCreating   0         0s
nginx-ingress-controller-pnch3   1/1       Running   0         5s
nginx-ingress-controller-pnch3   1/1       Running   0         6s
```

**Comment by gyliu513** [comment]
Can you please remove the `hostPort:` and re-test the case, I think it will be failed.

**Comment by roffe** [comment]
Removed whole port section. no error
```
➜  nginx-ingress git:(master) ✗ cat daemonset.yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: nginx-ingress-controller
  namespace: ingress-controller
spec:
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 1
  template:
    metadata:
      labels:
        name: nginx-ingress-controller
    spec:
      serviceAccountName: ingress-controller
      nodeSelector:
        ingress: external
      hostNetwork: true
      containers:
      - image: nginxdemos/nginx-ingress:1.0.0-alpine
        imagePullPolicy: Always
        name: nginx-ingress
        env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        # Uncomment the lines below to enable extensive logging and/or customization of
        # NGINX configuration with configmaps
        args:
         #- -v=3
          - -nginx-configmaps=$(POD_NAMESPACE)/nginx-config
          - -default-server-tls-secret=$(POD_NAMESPACE)/default-server-secret
```

```
➜  nginx-ingress git:(master) ✗ kc get po --namespace ingress-controller --watch
NAME                             READY     STATUS    RESTARTS   AGE
nginx-ingress-controller-5n751   1/1       Running   0          34m
nginx-ingress-controller-f75jr   1/1       Running   0          33m
nginx-ingress-controller-pnch3   1/1       Running   0          33m
nginx-ingress-controller-5n751   1/1       Terminating   0         34m
nginx-ingress-controller-5n751   0/1       Terminating   0         34m
nginx-ingress-controller-5n751   0/1       Terminating   0         34m
nginx-ingress-controller-5n751   0/1       Terminating   0         34m
nginx-ingress-controller-wwvx5   0/1       Pending   0         0s
nginx-ingress-controller-wwvx5   0/1       ContainerCreating   0         0s
nginx-ingress-controller-wwvx5   1/1       Running   0         2s
nginx-ingress-controller-pnch3   1/1       Terminating   0         34m
nginx-ingress-controller-pnch3   0/1       Terminating   0         34m
nginx-ingress-controller-wwvx5   1/1       Running   0         3s
nginx-ingress-controller-pnch3   0/1       Terminating   0         34m
nginx-ingress-controller-pnch3   0/1       Terminating   0         34m
nginx-ingress-controller-8t77c   0/1       Pending   0         0s
nginx-ingress-controller-8t77c   0/1       ContainerCreating   0         0s
nginx-ingress-controller-8t77c   1/1       Running   0         2s
nginx-ingress-controller-f75jr   1/1       Terminating   0         34m
nginx-ingress-controller-f75jr   0/1       Terminating   0         34m
nginx-ingress-controller-8t77c   1/1       Running   0         3s
nginx-ingress-controller-f75jr   0/1       Terminating   0         34m
nginx-ingress-controller-f75jr   0/1       Terminating   0         34m
nginx-ingress-controller-lg083   0/1       Pending   0         0s
nginx-ingress-controller-lg083   0/1       ContainerCreating   0         0s
nginx-ingress-controller-lg083   1/1       Running   0         2s
nginx-ingress-controller-lg083   1/1       Running   0         3s
```

**Comment by gyliu513** [comment]
@roffe How about left `containerPort` but only remove `hostPort`?

**Comment by roffe** [comment]
if i remove host port and apply the daemonset again it does not trigger a update of the pods.

I suspect this is due to it running on hostNetwork so the port declarations doesn't really matter

**Comment by x1957** [comment]
I had the port conflict problem.

env:
```
hostnetwork
nodeSelector to special node
```

I think same to daemonset

**Comment by hzxuzhonghu** [comment]
In theory if a daemonset use host network, and listen on port xxx. When update daemonset image, conflict must occur, because currently will create a new pod before shut down the old.
I remember several issues have been raised before.

**Comment by roffe** [comment]
@hzxuzhonghu i'm seeing the oposiet behaviour. It terminates first before creating new.

**Comment by roffe** [comment]
Tried reading through the release logs but can't find anything that the behaviour should have changed between 1.7.3 & 1.7.6

**Comment by roffe** [comment]
@gyliu513 you were on 1.7.3 if i don't remember wrong from slack conversation?

**Comment by crassirostris** [comment]
@roffe Add a non-zero terminationGracePeriodSeconds

**Comment by gyliu513** [comment]
@roffe yes, but I think that both 1.7 and above versions should all have such issues, so that's why I was requesting we should add a new strategy update for daemonset.

**Comment by ericchiang** [comment]
cc @diegs @rphillips

**Comment by diegs** [comment]
Sounds like this might be https://github.com/kubernetes/kubernetes/issues/50477 which was introduced sometime in 1.7 and fixed for 1.8. cc @janetkuo to confirm.

**Comment by gyliu513** [comment]
Thanks @diegs, I think that this is fixed in 1.8, close this for now.
